docs/proposed/config-split.md

```diff
@@ -0,0 +1,471 @@
+#define MBEDTLS_ECP_DP_BP512R1_ENABLED
+#define MBEDTLS_ECP_DP_CURVE25519_ENABLED
+#define MBEDTLS_ECP_DP_CURVE448_ENABLED
+#define MBEDTLS_ECP_DP_SECP192K1_ENABLED
+#define MBEDTLS_ECP_DP_SECP192R1_ENABLED
+#define MBEDTLS_ECP_DP_SECP224K1_ENABLED
+#define MBEDTLS_ECP_DP_SECP224R1_ENABLED
+#define MBEDTLS_ECP_DP_SECP256K1_ENABLED
+#define MBEDTLS_ECP_DP_SECP256R1_ENABLED
+#define MBEDTLS_ECP_DP_SECP384R1_ENABLED
+#define MBEDTLS_ECP_DP_SECP521R1_ENABLED
+#define MBEDTLS_ECDSA_C
+#define MBEDTLS_ECDSA_DETERMINISTIC
+#define MBEDTLS_ECJPAKE_C
+#define MBEDTLS_GCM_C
+#define MBEDTLS_GENPRIME
+#define MBEDTLS_HKDF_C
+#define MBEDTLS_HMAC_DRBG_C
+#define MBEDTLS_MD5_C
+#define MBEDTLS_PADLOCK_C
+#define MBEDTLS_PKCS1_V15
+#define MBEDTLS_PKCS1_V21
+#define MBEDTLS_POLY1305_C
+//#define MBEDTLS_PSA_CRYPTO_CONFIG
+//#define MBEDTLS_PSA_CRYPTO_SE_C
+#define MBEDTLS_RIPEMD160_C
+#define MBEDTLS_RSA_C
+#define MBEDTLS_SHA1_C
+#define MBEDTLS_SHA224_C
+#define MBEDTLS_SHA256_C
+#define MBEDTLS_SHA384_C
+#define MBEDTLS_SHA3_C
+#define MBEDTLS_SHA512_C
+```
+
+
+### In `mbedtls_config.h`, we have:
+#### SECTION Platform abstraction layer
+```
+#define MBEDTLS_NET_C
+//#define MBEDTLS_TIMING_ALT
+#define MBEDTLS_TIMING_C
+```
+
+
+#### SECTION General configuration options
+```
+//#define MBEDTLS_CONFIG_FILE "mbedtls/mbedtls_config.h"
+//#define MBEDTLS_USER_CONFIG_FILE "/dev/null"
+```
+
+
+#### SECTION TLS feature selection
+```
+#define MBEDTLS_DEBUG_C
+#define MBEDTLS_ERROR_C
+#define MBEDTLS_SSL_CACHE_C
+#define MBEDTLS_SSL_CLI_C
+#define MBEDTLS_SSL_COOKIE_C
+#define MBEDTLS_SSL_SRV_C
+#define MBEDTLS_SSL_TICKET_C
+#define MBEDTLS_SSL_TLS_C
+
+//#define MBEDTLS_PSK_MAX_LEN               32
+//#define MBEDTLS_SSL_CACHE_DEFAULT_MAX_ENTRIES      50
+//#define MBEDTLS_SSL_CACHE_DEFAULT_TIMEOUT       86400
+//#define MBEDTLS_SSL_CID_IN_LEN_MAX 32
+//#define MBEDTLS_SSL_CID_OUT_LEN_MAX 32
+//#define MBEDTLS_SSL_CID_TLS1_3_PADDING_GRANULARITY 16
+//#define MBEDTLS_SSL_CIPHERSUITES MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+//#define MBEDTLS_SSL_COOKIE_TIMEOUT        60
+//#define MBEDTLS_SSL_DTLS_MAX_BUFFERING             32768
+//#define MBEDTLS_SSL_IN_CONTENT_LEN              16384
+//#define MBEDTLS_SSL_MAX_EARLY_DATA_SIZE        1024
+//#define MBEDTLS_SSL_OUT_CONTENT_LEN             16384
+//#define MBEDTLS_SSL_TLS1_3_DEFAULT_NEW_SESSION_TICKETS 1
+//#define MBEDTLS_SSL_TLS1_3_TICKET_AGE_TOLERANCE 6000
+//#define MBEDTLS_SSL_TLS1_3_TICKET_NONCE_LENGTH 32
+```
+
+
+#### SECTION X.509 feature selection
+```
+#define MBEDTLS_PKCS7_C
+#define MBEDTLS_X509_CREATE_C
+#define MBEDTLS_X509_CRL_PARSE_C
+#define MBEDTLS_X509_CRT_PARSE_C
+#define MBEDTLS_X509_CRT_WRITE_C
+#define MBEDTLS_X509_CSR_PARSE_C
+#define MBEDTLS_X509_CSR_WRITE_C
+#define MBEDTLS_X509_USE_C
+
+//#define MBEDTLS_X509_MAX_FILE_PATH_LEN     512
+//#define MBEDTLS_X509_MAX_INTERMEDIATE_CA   8
+```
+
+
+#### SECTION Mbed TLS feature support
```

Comment:
I don't understand this section. IMO everything here should be moved to one of the three sections above.

Looks like the sections above only have boolean options ending in `_C` and non-boolean option, and this section got the rests, ie boolean options not ending in `_C`. As I've said before, I think having some options named with `_C` and others not is really throwing implementation details* into the user's face, which is bad design. Separating options into sections based on this is IMO just as bad.

*Actually not even implementation details, historical details. Currently TLS files are organized as `ssl_version_side.c` mostly but for some reason "side" is a `_C` option (`CLI_C` vs `SRV_C`) while "version" is not (`PROTO_TLS1_x`). Regardless, we're exposing users to things that they shouldn't need to know.

As a user, "I want support for TLS clients" is extremely similar to "I want support for TLS 1.3" and there is no reason for the two of them to live in different categories.

Specifically, I think:
- `MBEDTLS_ERROR_STRERROR_DUMMY`, `MBEDTLS_VERSION_xxx` should go in "General configuration options"
- the `MBEDTLS_X509` option should go to X.509
- everything else to TLS.

Not a blocker for the split as moving something internally between sections of the same file can be done later.

Reply:
Yes I thought about merging "Mbed TLS feature support" while working on the commit Re-organize "Mbed TLS modules" and "Module configuration options . I am happy to reorganize further `mbedtls_config.h` post-split. @gilles-peskine-arm are you okay with that?